src/ripple/shamap/impl/SHAMap.cpp

```diff
@@ -718,7 +718,7 @@ SHAMap::delItem(uint256 const& id)
         stack.pop();
 
         node = unshareNode(std::move(node), nodeID);
-        node->setChild(selectBranch(nodeID, id), prevNode);
+        node->setChild(selectBranch(nodeID, id), std::move(prevNode));
```

Comment:
If the copy that we did before was fine when `prevNode` was  holding a nullptr, moving the shared_ptr  should be fine as well.

Reply:
This looks OK to me. The behavior would be different if the loop could ever continue after `nodeID.isRoot()` is true - in that case you would enter the loop again with `prevNode` always empty in the new code, and it wouldn't be empty in the old code. However, I don't think that can ever happen - so I think this works as intended.

Reply:
Oh yes, thank you @seelabs and @nbougalis, I had misunderstood Nik's original objection.

Reply:
That's not the (potential!) problem, and I think my argument about the `use_count` added confusion. 

The existing code simply increases the reference count, whereas the _new_ code unconditionally moves out of `prevNode`.

Imagine if `nodeID.isRoot()` and `!stack.empty()` are both `true`: this changes the semantics of this code; whereas previously, _after_ the `setChild` you'd re-enter the loop with `prevNode` pointing to something, _now_ it'll point to `nullptr` because it's been moved out of.

Again, I am not convinced that's a concern in practice; I need to sit down and think about this code closely and carefully. But at the very least, we should acknowledge that this is a change in semantics that we must convince ourselves isn't problematic.

Reply:
I think that if `nodeID.isRoot()` is true, it means that we have iterated to the top of the tree, that `stack.empty()` will be true as well.

Reply:
Having spent some time thinking about the code, I agree that this should be OK and am removing my veto.

Reply:
Thanks @nbougalis , so marking the conversation as resolved.